### ISSUE TITLE
Non-ascii headers failure in 2.7.0

### ISSUE THREAD

**Issue by fernandomm** [mikel/mail#1166]
I'm running ruby 2.4.1p111 with Rails 5.1.4 and mail 2.7.0.

I noticed that non-ascii headers are no longer parsed and results in the message not being parsed at all.

If I try to parse a message with the header below, it will fail:

```
Subject: Resposta Automática
```

mail.subject will be nil and mail.headers will be empty, even if I have other headers.

The same message works fine in mail 2.6.6.

I compared both versions and started to test each commit. The issue starts at 7996f06f27e73e76f644041f79853b37e7b022a3 but works fine in 2b76d0a349895a84f02f1f86f762a4906bd133e8.

Based on the commit description, 7996f06f27e73e76f644041f79853b37e7b022a3 was a removal of old code to support older ruby versions. Is this a bug or just how mail handle things now?

**Comment by jeremy** [comment]
```ruby
>> Mail.new("Subject: Resposta Automática").subject
=> "Resposta Automática"
```

Could you paste a reproduction, with all steps included?

**Comment by fernandomm** [comment]
Sorry, the example wasn't complete. It seems that the parser is stopping at the From header before the subject. Although what causes the problem is the Subject header value.

Here is an example in rails console with "Subject: Resposta Automática":

```
[1] pry(main)> example = "From: \"Example.com\" <example@example.com>\nSubject: Resposta Automática\n\nExample"
=> "From: \"Example.com\" <example@example.com>\nSubject: Resposta Automática\n\nExample"
[2] pry(main)> Mail.new(example).errors
=> [["From",
  "\"Example.com\" <example@example.com>\nSubject: Resposta Autom\xC3\xA1tica\n\nExample",
  #<Mail::Field::IncompleteParseError: Mail::AddressList can not parse |"Example.com" <example@example.com>
Subject: Resposta Automática

Example|: Only able to parse up to "\"Example.com\" <example@example.com>">]]
```

Here is an example with "Subject: Example" ( only ascii chars ). It works fine:

```
[3] pry(main)> example = "From: \"Example.com\" <example@example.com>\nSubject: Example\n\nExample"
=> "From: \"Example.com\" <example@example.com>\nSubject: Example\n\nExample"
[4] pry(main)> Mail.new(example).errors
=> []
```

**Comment by jeremy** [comment]
```
>> Mail.new(Mail::Utilities.binary_unsafe_to_crlf("From: \"Example.com\" <example@example.com>\nSubject: Resposta Automática\n\nExample")).subject
=> "Resposta Automática"
```

It's because the email has LF line endings, not CRLF, and we can't detect that it's safe to automatically convert them LF → CRLF.

Highly recommend switching to CRLF line endings if you can. This mostly affects test fixtures that have been saved with UNIX line endings (`\n`) instead of DOS line endings (`\r\n`). CRLF is required, but Mail has long attempted to support LF for convenience. But that results in sometimes mangling binary values, no good.

See #1159 and #1160 for more.

**Comment by rubys** [comment]
FWIW, what this change breaks is not a test case, but code deployed in production:

https://github.com/apache/whimsy/blob/master/www/secretary/workbench/views/actions/icla.json.rb#L223

TL;DR: this affects mail constructed from erb files.

What is particularly odd is that with this change. #to returns back a String in this case instead of an Array.  At a minimum, I would suggest that that be fixed.

**Comment by jeremy** [comment]
Ah cool! Haven't seen that usage. Most other cases of parsing with non-RFC line endings are developers' text fixtures 😊

> What is particularly odd is that with this change. #to returns back a String in this case instead of an Array. At a minimum, I would suggest that that be fixed.

Got a before/after example of what broke?

**Comment by rubys** [comment]
test2 is representative of the breakage.

Here is the original report: https://lists.apache.org/thread.html/788fe70423676c4636fd4c6e3d65d971124a7c620175af93e9910040@%3Cdev.whimsical.apache.org%3E

Note the symptom: 

    #<NoMethodError: undefined method `map' for #<String:0x007f9ba34add28> Did you mean? tap>

**Comment by jeremy** [comment]
I see. That happens when parsing fails, resulting in an UnstructuredField fallback, resulting in an apparent string address rather than an array of parsed addresses. And parsing fails because of the LF line endings. Same deal.

**Comment by jeremy** [comment]
@rubys, you can fix immediately by converting the template results to CRLF before parsing:
```ruby
mail = Mail.new(Mail::Utilities.binary_unsafe_to_crlf(template('acreq.erb')))
```

**Comment by rubys** [comment]
If this were the only place where I parsed an email message, that would be fine.  Meanwhile, I'm really tempted to monkey patch `safe_for_line_ending_conversion?` as follows:

    string.encoding == Encoding::BINARY or not string.valid_encoding?

**Comment by jeremy** [comment]
@rubys That's essentially the heuristic change we'll do for Ruby >= 1.9. PR welcome if you'd like to pursue it. 👍

**Comment by jeremy** [comment]
Fixed by #1168
